clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -253,11 +254,20 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
         SemaRef.Diag(e->getExprLoc(), diag::err_builtin_target_unsupported)
             << Name << "SYCL device";
       }
+
+      // Disallow functions with neither definition nor SYCL_EXTERNAL mark
+      // Only validate really called methods
+      if (SYCLCG.getNode(Callee) &&
```

Comment:
Are the following included in SYCLCG (sorry I'm not familiar with that code) if they aren't (directly or indirectly) called by a kernel in this TU?
- a function is called by a function marked as `SYCL_EXTERNAL`
- a function whose device function pointer is taken

I'm asking because both of those types of functions can indirection be called from a SYCL kernel and should also trigger the warning, correct?

Reply:
There is a test for function pointers: sycl/test/function-pointers/fp-as-kernel-arg.cpp et al. It compiles and runs OK.

Reply:
> Are the following included in SYCLCG (sorry I'm not familiar with that code) if they aren't (directly or indirectly) called by a kernel in this TU?
> * a function is called by a function marked as `SYCL_EXTERNAL`

Yes.
> * a function whose device function pointer is taken

Yes, if it's marked with `device_indirectly_callable` attribute.

Reply:
> There is a test for function pointers: sycl/test/function-pointers/fp-as-kernel-arg.cpp et al. It compiles and runs OK.

I wasn't worried that this change would break something (false positive). I was worried whether the added diagnostic would properly work for these types of functions (false negative). If one calls a undefined function from the `device_indirectly_callable` or `SYCL_EXTERNAL` function then the diagnostic should work too. I don't see a test for that. Not sure whether we need one. Given the answer the functions are added, everything should be fine.

Reply:
It turned out that callgraph contains only functions with definitions.